resources/ansible/roles/downloaders/files/download.sh

```diff
@@ -0,0 +1,52 @@
+#!/usr/bin/env bash
+
+CONTACT_PEER="${1:-}"
+
+CONTACT_PEER_ARG=""
+if [[ -n "$CONTACT_PEER" ]]; then
+  CONTACT_PEER_ARG="--peer $CONTACT_PEER"
+fi
+
+if ! command -v safe &> /dev/null; then
+  echo "Error: 'safe' not found in PATH."
+  exit 1
+fi
+
+LOG_FILE="./uploaded_files.log"
+DOWNLOAD_DIR="./downloaded_files"
+SLEEP_INTERVAL=5
+
+mkdir -p "$DOWNLOAD_DIR"
+
+download_file() {
+  local file_ref=$1
+  # Multiple downloaders can be running on the same machine, so one or more
+  # could select the same file address at the same time. We therefore use a GUID
+  # to ensure a unique output file.
+  local output_path="$(uuidgen).dat"
+  (
+    cd "$DOWNLOAD_DIR"
+    echo "Downloading file: $file_ref"
+    safe $CONTACT_PEER_ARG files download "$output_path" "$file_ref"
+    if [[ $? -eq 0 ]]; then
+      echo "Downloaded $file_ref to $output_path"
+    else
```

Comment:
I know we don't have much time but if we want telegraf to monitor download performance we could add something like:

```
  echo "$time,$file_size_kb" >> "./downloader_metrics.csv"
```

The code snippet doesn't include the header for the 2 columns but yeah.

This could allow us to have telegraf record each download attempt (time it took), and size it download on disk (whether its 0 or > 0) and get that graphed.

Not a blocker for tomorrow, but maybe something to circle back with in a few days?

Reply:
Hey, thanks for this, but I think I'll just leave it. We can easily add this in when we actually have the monitoring.